cgfs.py

```diff
@@ -160,34 +870,205 @@ def truncate(self, length):
+
+    def get_feedback(self, payload):
+        f_name = self.cgfs.strippath(payload['file'])
+
+        with self.cgfs._lock:
+            try:
+                f = self.cgfs.get_file(f_name, expect_type=SingleFile)
+            except:
+                return {'ok': False, 'error': 'File not found', 'f': f_name}
+
+            if isinstance(f, TempFile):
+                return {'ok': False, 'error': 'File not a sever file'}
+
+            try:
+                res = cgapi.get_feedback(f.id)
+            except:
+                return {'ok': False, 'error': 'The server returned an error'}
+
+            return {'ok': True, 'data': res}
+
+    def add_feedback(self, payload):
+        f_name = self.cgfs.strippath(payload['file'])
+        line = payload['line']
+        message = payload['message']
+
+        with self.cgfs._lock:
+            try:
+                f = self.cgfs.get_file(f_name, expect_type=SingleFile)
+            except:
+                return {'ok': False, 'error': 'File not found'}
+
+            if isinstance(f, TempFile):
+                return {'ok': False, 'error': 'File not a sever file'}
+
+            try:
+                cgapi.add_feedback(f.id, line, message)
+            except:
+                return {'ok': False, 'error': 'The server returned an error'}
+
+            return {'ok': True}
+
+
 class CGFS(LoggingMixIn, Operations):
-    def __init__(self, latest_only):
+    API_FD = 0
+
+    def __init__(
+        self,
+        latest_only,
+        socketfile,
+        mountpoint,
+        fixed=False,
+        tmpdir=None,
+        rubric_append_only=True
+    ):
         self.latest_only = latest_only
+        self.fixed = fixed
         self.files = {}
-        self.fd = 0
+        self.fd = 1
+        self.mountpoint = mountpoint
+        self._lock = threading.RLock()
+        self._open_files = {}
+
+        self._tmpdir = tmpdir
+
+        self._socketfile = socketfile
+        self.socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+        self.socket.bind(self._socketfile)
+        self.socket.listen()
+        self.api_handler = APIHandler(self)
+        threading.Thread(
+            target=self.api_handler.run, args=(self.socket, )
+        ).start()
+        self.special_socketfile = SocketFile(
+            bytes(socketfile, 'utf8'), '.api.socket'
+        )
+
+        self.rubric_append_only = rubric_append_only
 
         self.files = Directory(
             {
                 'id': None,
                 'name': 'root'
             }, type=DirTypes.FSROOT
         )
-        self.files.getattr()
-        self.load_courses()
+
+        with self._lock:
+            self.files.getattr()
+            self.files.insert(self.special_socketfile)
+            self.files.insert(
+                SpecialFile(
+                    '.cg-mode', b'FIXED\n' if self.fixed else b'NOT_FIXED\n'
+                )
+            )
+            self.load_courses()
+        print('Mounted')
```

Comment:
@olmokramer wrote:

> This should only be printed in `--verbose` mode in my opinion. I would probably run the filesystem in the background (with `cgfs user mountpoint &`) and it's very annoying when backgrounded programs print to stdout.

@libre-man wrote:

> No this should only not be printed in `--daemon` mode imo.

That would be fine. Couldn't find a `--daemon` mode, though?

Reply:
No that is expected as it is not a valid flag at the moment.

Reply:
Ok fine